File: leet-code/92. Reverse Linked List II/ReverseLinkedListII92.java
public class ReverseLinkedListII92 {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if (head.next == null) {
            return head;
        }

        ListNode root = new ListNode(501);
        root.next = head;

        ListNode node = root;
        for (int i = 1; i < left; i++) {
            node = node.next;
        }
        ListNode leftNode = node;

        for (int i = left; i <= right; i++) {
            node = node.next;
        }
        ListNode rightNode = node.next;
        // 다음 노드와의 연결을 끊는다
        node.next = null;

        ListNode reversedNode = reverse(leftNode.next);
        leftNode.next = reversedNode;
        while (reversedNode.next != null) {
            reversedNode = reversedNode.next;
        }
        reversedNode.next = rightNode;
        return root.next;
    }

    private ListNode reverse(ListNode node) {
        ListNode prev = null;
        while (node != null) {
            ListNode next = node.next;
            node.next = prev;
            prev = node;
            node = next;
        }
        return prev;
    }
}

File: leet-code/92. Reverse Linked List II/ReverseLinkedListII92.kt
package com.example

class ReverseLinkedListII92 {
    fun reverseBetween(head: ListNode?, left: Int, right: Int): ListNode? {
        val root = ListNode(0)
        root.next = head
        var pre = root
        var cur = head

        for (i in 1..<left) {
            pre = cur!!
            cur = cur.next
        }

        val curRef = cur
        var prev: ListNode? = null
        for (i in left until right + 1) {
            val next = cur!!.next
            cur.next = prev
            prev = cur
            cur = next
        }

        pre.next = prev
        curRef!!.next = cur
        return root.next
    }
}
